File: 0022-generate-parentheses/0022-generate-parentheses.java
class Solution {
    public List<String> generateParenthesis(int n) {
        List<String> list = new ArrayList<>();
        generateValidParenthesis(list, n, n, "");
        return list;
    }

    private void generateValidParenthesis(List<String> list, int open, int close, String current) {

        // Base case
        if (open == 0 && close == 0) {
            list.add(current);
            return;
        }

        // Place an opening parenthesis
        if (open > 0) {
            generateValidParenthesis(list, open - 1, close, current + "(");
        }

        // Place a closing parenthesis only if it keeps the string valid
        if (close > open) {
            generateValidParenthesis(list, open, close - 1, current + ")");
        }
    }
}

File: 0022-generate-parentheses/0022-generate-parentheses.js
/**
 * @param {number} n
 * @return {string[]}
 */
var generateParenthesis = function(n) {
    var list = [];
    generateValidParenthesis(list,n,n,"");
    return list; 
};

function generateValidParenthesis(list,open,close,str){
    if(open == 0 && close == 0){
        list.push(str);
        return;
    }

    if(open>0){
        generateValidParenthesis(list,open-1,close,str+'(');
    }

    if(close > open){
        generateValidParenthesis(list,open,close-1,str+')');
    }
}